hips/hip-9999.md

```diff
@@ -0,0 +1,76 @@
+---
+hip: 9999
+title: "Document backwards-compatibility rules"
+authors: [ "Marc Khouzam <marc.khouzam@montreal.ca>" ]
+created: "2020-09-18"
+type: "process"
+status: "draft"
+---
+
+## Abstract
+
+Define and document the backwards-compatibility rules that apply to Helm minor and patch releases.
+
+## Motivation
+
+Helm aims to strictly respect [Semantic Versioning][semver], where minor and patch releases must be 100% backwards-compatible.
+Backwards-compatibility implies that when moving to a new backwards-compatible release:
+* tools (such as Continuous Integration tools) using the helm CLI will not require any changes
+* programs using the Helm Go API will continue to compile 
+
+Some contributors to the Helm code base may not be comfortable with the details of what constitutes a compatibility-breaking change.
+Furthermore, in some cases, there can be a difference of opinion, even amongst the maintainers, on what constitutes such a change.
+
+Having a well-defined, documented set of backwards-compatibility rules for the Helm code-base will allow to:
+* avoid compatibility-breaking changes being merged by mistake
+* guide contributors in understanding helm's backwards-compatibility requirements
+* remove any uncertainty when it comes to enforcing backwards compatibility
+* provide a basis for discussion and possible updates to how helm handles backwards compatibility
+
+## Rationale
+
+The maintainers have agreed that a formal definition of Helm's backwards-compatibility rules would be beneficial.
+
+## Specification
+
+The following compatibility rules will apply to helm minor and patch releases:
+* exported Go APIs will remain compatible as per this [article][go-module-comp]
+* CLI commands:
+    * are case-sensitive
+    * must not be removed
+    * must not be renamed
+    * must not be moved elsewhere in the command hierarchy
+* CLI flags:
+    * are case-sensitive
+    * must not be removed
+    * must not be renamed
+    * must not be moved in the command hierarchy
+    * must not change type
+* CLI output must:
+    * ???
+
+## Security implications
+
+N/A
```

Comment:
Compatibility rules may be broken or bent in the name of security. For example, if a flag once took an int64, but it was determined that doing so led to a buffer overflow, the datatype of that parameter could be changed.

In some cases, entire functions, template commands, etc. could be altered or removed if deemed absolutely necessary. Files formats can change as well. Of course, it should go without saying that we as maintainers should do our best to not have this happen if at all possible. This caveat is mainly to inform the end users that in extreme cases, we will break compatibility.

Reply:
Are you familiar with what other projects do in such cases?  Do such changes technically require a major release?  I know it sound excessive, but if a project will no longer compile when moving to a new minor release, that sounds contradictory to me.

I've added your text to the Security section, and will add my question as an open issue until we have an answer.